### ISSUE TITLE
Two new feature request for vehicle events

### ISSUE THREAD

**Issue by umitemre** [multitheftauto/mtasa-blue#1445]
I actually have two feature requests. 

- First is `onPlayerQuit` event should trigger `onVehicleExit` and `onPlayerVehicleExit`, `onVehicleStartExit`  events. 

- Second is none of the events gets triggered at the right time. For example if a vehicle is in high speed and player pressed ENTER or F key to leave vehicle `onVehicleStartExit` gets triggered. That's OK but `onPlayerVehicleExit` and `onVehicleExit` don't get triggered until the ped rolled a few meters and get back on his feet again.

**Is your feature request related to a problem? Please describe.**
When a player quits server while in a vehicle, we have to some extra work on the vehicle player was before quitting. I can not detect it since `onPlayerVehicleExit` and `onVehicleExit` events are not being triggered. I triggered these events on manually but `getVehicleOccupants` returns the player is about to leave. It shouldn't be there because when player is quitting, he should be considered as exited from vehicle as he quits the server.

For second feature request, there are certain commands and binds when player can use in vehicle. This commands and binding still works before `onPlayerVehicleExit` gets called because `getPedOccupiedVehicle` still returns that vehicle even ped and the vehicle are seperate from each other. `onVehicleStartExit` event is also too soon to consider player exited the vehicle.

**Describe the solution you'd like**
`onPlayerQuit` should remove the player from vehicle he is in and sync it so script can detect he's on foot and not in that vehicle.

`getPedState()` can be added to describe player's current state: on foot, driver or passenger etc.. 
Even `getPedOccupiedVehicle` returns a valid vehicle, we can check if ped state is on foot.

**Comment by dmi7ry** [comment]
> * First is `onPlayerQuit` event should trigger `onVehicleExit` and `onPlayerVehicleExit`, `onVehicleStartExit`  events.

I don't think it should. You can just trigger the events in `onPlayerQuit`, if you need it.

**Comment by CrosRoad95** [comment]
> onPlayerQuit event should trigger onVehicleExit and onPlayerVehicleExit, onVehicleStartExit events.

it can ruin backwards compatibility

**Comment by umitemre** [comment]
> > * First is `onPlayerQuit` event should trigger `onVehicleExit` and `onPlayerVehicleExit`, `onVehicleStartExit`  events.
> 
> I don't think it should. You can just trigger the events in `onPlayerQuit`, if you need it.

As I mention in the original post, I've tried to trigger these events manually but `getVehicleOccupants()` returns the player who is quitting too. So I have to manually filter and do an extra for loop.



> > onPlayerQuit event should trigger onVehicleExit and onPlayerVehicleExit, onVehicleStartExit events.
> 
> it can ruin backwards compatibility

Can you explain further on how this can ruin it? 

Also this was my suggestion and this can be archived by doing or adding any other thing that doesn't ruin backwards compatibility.

**Comment by qaisjp** [comment]
The player _is_ inside the vehicle when the player is quitting - they haven't left the vehicle / aren't on foot, so that's why `getVehicleOccupants` includes that player.

I'd suggest refactoring your code and making a "cleanup" function that is called when they leave the vehicle, or when they quit server.

Other edge cases you need to consider (which would also need to use that cleanup function):

- `onVehicleStartExit` is not called when `removePedFromVehicle` is used
- None of the vehicle exit events are called when `destroyElement(vehicle)` is used

**Comment by ghost** [comment]
> First is onPlayerQuit event should trigger onVehicleExit and onPlayerVehicleExit, onVehicleStartExit events.

Your reasoning is not convincing since this will add extra overhead and there's no need to because you can detect the vehicle occupants by calling other functions in onPlayerQuit event. 

> Second is none of the events gets triggered at the right time. For example if a vehicle is in high speed and player pressed ENTER or F key to leave vehicle onVehicleStartExit gets triggered. That's OK but onPlayerVehicleExit and onVehicleExit don't get triggered until the ped rolled a few meters and get back on his feet again.

> getPedState() can be added to describe player's current state: on foot, driver or passenger etc..
Even getPedOccupiedVehicle returns a valid vehicle, we can check if ped state is on foot.

You can use [getPedTask](https://wiki.multitheftauto.com/wiki/GetPedTask) for that. Check for "TASK_COMPLEX_LEAVE_CAR" primary task at index 3. 

> I triggered these events on manually but getVehicleOccupants returns the player is about to leave. It shouldn't be there because when player is quitting, he should be considered as exited from vehicle as he quits the server.

That's how it should be because the player is in the vehicle while quitting the server. As @qaisjp suggested, you should consider refactoring your code. This looks like a scripting problem.

**Comment by umitemre** [comment]
> > First is onPlayerQuit event should trigger onVehicleExit and onPlayerVehicleExit, onVehicleStartExit events.
> 
> Your reasoning is not convincing since this will add extra overhead and there's no need to because you can detect the vehicle occupants by calling other functions in onPlayerQuit event.
> 
> > Second is none of the events gets triggered at the right time. For example if a vehicle is in high speed and player pressed ENTER or F key to leave vehicle onVehicleStartExit gets triggered. That's OK but onPlayerVehicleExit and onVehicleExit don't get triggered until the ped rolled a few meters and get back on his feet again.
> 
> > getPedState() can be added to describe player's current state: on foot, driver or passenger etc..
> > Even getPedOccupiedVehicle returns a valid vehicle, we can check if ped state is on foot.
> 
> You can use [getPedTask](https://wiki.multitheftauto.com/wiki/GetPedTask) for that. Check for "TASK_COMPLEX_LEAVE_CAR" primary task at index 3.
> 
> > I triggered these events on manually but getVehicleOccupants returns the player is about to leave. It shouldn't be there because when player is quitting, he should be considered as exited from vehicle as he quits the server.
> 
> That's how it should be because the player is in the vehicle while quitting the server. As @qaisjp suggested, you should consider refactoring your code. This looks like a scripting problem.

OK. 

By the way the main reason why I am suggesting the first suggestion was to properly respawn idle vehicles.

`toggleVehicleRespawn()` works in cases like explosion, being idle etc. (even for trailers, too) but when a vehicle parked near where it is created after used, it doesn't respawn. This could be a bug or not. If it's not then I'm gonna refactor my code.

**Comment by qaisjp** [comment]
> `toggleVehicleRespawn()` works in my cases explosion, being idle etc. but when a vehicle parked near where it is created after used, it doesn't respawn. This could be a bug or not. If it's not then I'm gonna refactor my code.

Interestingly, a few days ago I hopped into a car created by the `play` resource and just sat there for a few minutes. Then got stuck by a new vehicle respawning in the same place (the same model).

**Comment by umitemre** [comment]
> > `toggleVehicleRespawn()` works in my cases explosion, being idle etc. but when a vehicle parked near where it is created after used, it doesn't respawn. This could be a bug or not. If it's not then I'm gonna refactor my code.
> 
> Interestingly, a few days ago I hopped into a car created by the `play `resource and just sat there for a few minutes. Then got stuck by a new vehicle respawning in the same place (the same model).

In addition to that, sometimes bikes can fell over when they have respawned. Specially when they already fell over (try to jump out of bike while going in a little bit fast)

**Comment by qaisjp** [comment]
What do you mean by 'bikes can fell off'?

**Comment by umitemre** [comment]
> What do you mean by 'bikes can fell off'?

It was going be "fall over". Could be dropped too. I mean not standing on their two wheels, lying on the ground. Here's a little GIF that I've made on WhatsApp 🙃 

![ezgif com-optimize](https://user-images.githubusercontent.com/9094961/82323552-e613c200-99e0-11ea-9f7b-c9a324df0542.gif)

NOTE: I think when the bikes not streamed to any player that respawn fine without falling over. I've teleported somewhere else to respawn all vehicles and came back and those two bikes were standing still.

**Comment by qaisjp** [comment]
It might be falling over because when it respawns it is not frozen or damageproof. I believe the play resources freezes and makes them damage proof until you get inside or something?

**Comment by umitemre** [comment]
> It might be falling over because when it respawns it is not frozen or damageproof. I believe the play resources freezes and makes them damage proof until you get inside or something?

Not freezing. When user leaves the vehicle or when it's respawned then it gets damage proof but not get frozen.
